### PR TITLE
Feature/sbachmei/mic 3768 create households pipeline

### PR DESCRIPTION
**Pull request by stevebachmeier**
### Description
- *Category*: Refactor
- *JIRA issue*: [MIC-3768](https://jira.ihme.washington.edu/browse/MIC-3768)
- *Research reference*: na

### Changes and notes
This PR is analogous to the business_details pipeline; it adds a new
pipeline that returns `address_id` and `housing_type` based on
each simulant's `household_id`. This PR does NOT yet undo
those returned columns from the population state table; that work
is for a followup PR.

All integration tests that previously relied on `address_id` or `housing_type`
in the state table are converted to instead using the new pipeline.

### Verification and Testing
all pytests that require the pipeline values pass.

### CODE REVIEW

File: src/vivarium_census_prl_synth_pop/components/household.py [hunk 1]
```diff
@@ -135,13 +148,41 @@ def on_time_step(self, event: Event):
...
         )
 
         self.population_view.update(pop)
+
+    ##################
+    # Helper methods #
+    ##################
+
+    def generate_initial_households(self, pop_data: SimulantData) -> pd.DataFrame():
+        households = self.population_view.subview(["household_id", "housing_type"]).get(
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/components/household.py:167]: Is there not a way to get housing type without looking at the housing_type column? If not, you're going to have trouble when you try to delete it. Can't we just define the GQ households explicitly and then set the rest of them as "Standard"?

File: src/vivarium_census_prl_synth_pop/components/household.py [hunk 2]
```diff
@@ -135,13 +148,41 @@ def on_time_step(self, event: Event):
...
+    def generate_household_details(self, idx: pd.Index) -> pd.DataFrame:
+        pop = self.population_view.subview(["household_id"]).get(idx)
+        household_details = pop.join(
+            self.households,
+            on="household_id",
+        )
+        # Set housing_type dtypes
+        household_details["housing_type"] = household_details["housing_type"].astype(
+            pd.CategoricalDtype(categories=HOUSING_TYPES)
+        )
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/components/household.py:186]: Is the dtype of `self.households["housing_type"]` not Categorical already? Does the merge not preserve the dtype?

File: src/vivarium_census_prl_synth_pop/components/person.py [hunk 3]
```diff
@@ -7,6 +9,7 @@
 from vivarium.framework.randomness import RESIDUAL_CHOICE
 from vivarium.framework.utilities import from_yearly
 
+from vivarium_census_prl_synth_pop.components import HouseholdMigration
```
Comment (rmudambi) [inline src/vivarium_census_prl_synth_pop/components/person.py:12]: Were you intending to type `self.household_migration` explicitly? I don't see where you use this import.

**Review by rmudambi — CHANGES_REQUESTED** [review]

File: src/vivarium_census_prl_synth_pop/components/household.py [hunk 4]
```diff
@@ -78,26 +86,31 @@ def on_initialize_simulants(self, pop_data: SimulantData) -> None:
...
             }
             households["address_id"] = households["household_id"].map(address_assignments)
             self.population_view.update(households)
-        else:
+        else:  # newborns
             parent_ids = pop_data.user_data["parent_ids"]
             mothers = self.population_view.get(parent_ids.unique())
+            mothers["address_id"] = mothers["address_id"].astype(
+                int
+            )  # FIXME - why are these floats??
```
[the remark this answers is not in the document]
Comment (stevebachmeier) [reply src/vivarium_census_prl_synth_pop/components/household.py:105]: Yeah, I'm just not immediately seeing where the nans are being introduced. I admittedly having looked very hard.

File: src/vivarium_census_prl_synth_pop/components/household.py (hunk 1 above)
Comment (stevebachmeier) [reply src/vivarium_census_prl_synth_pop/components/household.py:167]: Oh I was just generating a new `initialized_housing_type` column that was used here instead as well as in fertility for newborns. I like your idea more if it works in fertility.

File: src/vivarium_census_prl_synth_pop/components/household.py (hunk 2 above)
Comment (stevebachmeier) [reply src/vivarium_census_prl_synth_pop/components/household.py:186]: It wasn't coming in as such

File: src/vivarium_census_prl_synth_pop/components/person.py (hunk 3 above)
Comment (stevebachmeier) [reply src/vivarium_census_prl_synth_pop/components/person.py:12]: leftover from when I was sub-componenting it

File: integration_tests/test_domestic_migration.py [hunk 5]
```diff
@@ -105,40 +106,50 @@ def test_households_move(simulants_on_adjacent_timesteps):
...
+        ("household_id", "household_details.address_id"),
+    ],
+)
+def test_unit_address_uniqueness(populations, unit_id_col, address_id_col):
+    """Check that all units (households and businesses) have unique details and
+    also that simulants of the same unit have the same details
     """
-    _test_address_uniqueness(
+    breakpoint()
+    _test_unit_address_uniqueness(
```
Comment (rmudambi) [inline integration_tests/test_domestic_migration.py:144]: Remove breakpoint. Why did you retain the helper function?

File: integration_tests/test_domestic_migration.py [hunk 6]
```diff
@@ -159,18 +170,21 @@ def _test_address_uniqueness(pops, unit_id_col, address_id_col, other_address_co
...
 @pytest.mark.skip(reason="Wait for state/puma to be added (MIC-3728)")
-def test_addresses_during_moves(simulants_on_adjacent_timesteps):
-    """Check that address details change after a move. Note that
-    this test does not distinguish between the different types of moves (eg
-    household vs individual moves)
-    """
-    # business moves
+def test_addresses_during_moves(simulants_on_adjacent_timesteps, unit_id_col, address_id_col):
+    """Check that business address details change after a move."""
     _test_addresses_during_moves(
```
Comment (rmudambi) [inline integration_tests/test_domestic_migration.py:183]: Same as above

File: src/vivarium_census_prl_synth_pop/components/household.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rmudambi) [reply src/vivarium_census_prl_synth_pop/components/household.py]: It happens in `population_view.update`

**Review by rmudambi — APPROVED** [review]

**Pull request merged**